### ISSUE TITLE
Bug in git cli command line wrapper: SSH keys can't be unlocked any more

### ISSUE THREAD

**Issue by Binarus** [gitextensions/gitextensions#11379]
### Environment

- Git Extensions 4.2.0.17591
- Build bf814d4f1c6f178eacef38d592b700f9f2519dcb
- Git 2.42.0.windows.2
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.25
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions
```
    Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### Issue description

In previous versions, when we opened Git bash from within Git Extensions, and worked in Git bash with remote repositories via SSH, we were asked the passphrases for the respective SSH keys directly in the Git bash terminal. Once we had typed the passphrase into the terminal, or had copied it from a password manager to there, the respective SSH keys were unlocked, and the operation in question was carried out. That worked reliably.

Since the upgrade to the version indicated above, we can't type passphrases for the SSH keys directly in the terminal any more, or copy them to there from our password manager. Instead, a GUI dialog box opens where we should type (or copy) our passphrase into. Apart from the fact that this is worrying and superfluous, the real problem is that this git cli command stdin wrapper does not process our passphrases correctly.

That is, even if we type the passphrase correctly in that GUI dialog, the respective SSH key is not unlocked, claiming that the passphrase is wrong. Of course, this has disabled our access to all our remote repositories.

### Steps to reproduce

- Create a remote repository that can be reached via SSH.
- Protect the private SSH key with a long and complex passphrase that contains special characters (*).
- Open Git bash from within Git Extensions.
- From within Git bash, try to access the remote repository (e.g. clone it), which will make the Git cli stdin wrapper appear.
- Type your passphrase into the Git command line stdin wrapper.
- Observe that it claims that your passphrase is wrong and that the SSH key cannot be unlocked, although you have entered the passphrase correctly. This prevents you from accessing your remote repositories.

(*) I can't tell how the passphrase must be crafted to reproduce the problem. Unfortunately, I cannot provide the passphrases that I am using, because they are used in production. However, I *believe* that German umlauts in the passphrase are enough to trigger the problem.

### Did this work in previous version of GitExtensions?

Ambiguous answer: Yes and no.

Yes, *entering the passphrase* worked in all previous versions, because the previous versions did not feature that Git cli command line stdin wrapper. Typing or copying the passphrase into the Git bash terminal has always worked like a charm.

But no, *the command line stdin wrapper* did not work in previous versions, because it didn't exist yet (at least it wasn't active by default).

### Diagnostics

I apologize that I don't have the time to generate diagnostics or to answer further questions right now. However, I believe that the problem can easily be reproduced.

Finally, two hints:

- It is obvious that the stdin wrapper does not handle character encodings correctly. Either the conversion from the source encoding (Windows) to the internal encoding (if any) works incorrectly, or the conversion from the internal encoding to the Git bash encoding works incorrectly.
- If code from the Putty project has been used to implement that stdin wrapper: We had exactly the same problem with Putty's passphrase converter (which is the reason why we stopped using Putty a while ago).

Thank you very much,

Binarus

**Comment by mstv** [comment]
As a workaround, you might unset the environment variable `SSH_ASKPASS_REQUIRE`, if I got #11199 right.

Should we consider to set these variables for git execution only but not when launching shells?

**Comment by Binarus** [comment]
Well, although I don't see the need for that stdin wrapper, I suppose that there surely was a reason to introduce it. Therefore, probably the best solution would be to fix the bug that's in the wrapper.

Thank you very much for your tip about environment variables! That's exactly the thing we already had done before writing this bug report, thanks to another helpful soul that had drawn our attention to the SSH documentation and to that environment variable. `export SSH_ASKPASS_REQUIRE=never` in `.bashrc` was enough to restore the old behavior and to let the user enter their passphrases in the Git bash terminal again, as in previous versions, where it works reliably.

Regarding the Git execution vs. execution in the shell: I'll guest I can test in a few hours whether the git execution works or whether it is affected by the flaw as well.

**Comment by Binarus** [comment]
> Should we consider to set these variables for git execution only but not when launching shells?

That wouldn't solve the problem. As promised, I have tested what you probably mean by "git execution", and as I had suspected, it suffers exactly the same problem as the Git bash. That is:

- In Git Extensions, I have opened a repository that I had on my local dist and whose remote can be reached via SSH.
- In Git Extensions, I have chosen "Commands -> Pull/Fetch...", have adjusted the options as needed in the dialog box that appeared, and have clicked the "Fetch" button in that dialog box.
- The Git cli command line stdin wrapper opened and of course again didn't process the passphrase correctly.

So it doesn't matter whether you work from within Git Bash or use Git Extensions' UI. Wherever I have to enter one of my SSH passphrases, it fails.

The situation with Git Extension is the same as with Git bash: In previous versions of Git Extensions, when the SSH passphrase was needed, there was a mini console window where we could enter it directly. Now, the mini console window still appears, but we can't enter our passphrases there; instead, again, the Git command line stdin wrapper appears.

**Comment by SlugFiller** [comment]
Did you test in older versions of Git Extensions, whether the "mini console" even works for your password? Because it's not 100% compatible with `cmd` or `bash`, and tends to mess up input if you need to press any special keys (e.g. ctrl, alt, backspace, etc).

Also, looking at it more closely, the issue could actually be password length, rather than contents. While the wrapper has support for unicode (`TCHAR` is used everywhere), the password is [limited to 100 characters](https://github.com/gitextensions/gitextensions/blob/44be238d28e5c68ee11bca3f67351d7383777e65/GitExtSshAskPass/SshAskPass.cpp#L43). Could this be part of the issue?

As a test, try running the program directly from inside `cmd` or `bash`. It should print whatever you input into it, to the console. It should be as simple as opening `cmd`, `bash`, or `powershell` from the Git Extensions toolbar, and then running:
- `bash`: `"$SSH_ASKPASS"`
- `cmd`: `"%SSH_ASKPASS%" | more`
- `powershell`: `& $env:SSH_ASKPASS | more`

It should print the result to the console. If the result comes out garbled or truncated, it indicates there's a problem in the wrapper.

**Comment by pmiossec** [comment]
@Binarus 

> Well, although I don't see the need for that stdin wrapper, I suppose that there was a good reason to introduce it. Therefore, probably the best solution would be to fix the bug that's in the wrapper (or is it a feature of the underlying SSH utilities?).

Can you try this attempt of a fix #11381 ? Get the new `GitExtSshAskPass.exe` from the msi file (the portable version doesn't have it inside 😢 ) in artifacts: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/48585850/artifacts

Then uncompress the msi and get  `GitExtSshAskPass.exe` from there (7-zip is doing it well).

> Thank you very much for your tip about environment variables! That's exactly the thing we already had done before writing this bug report, thanks to another helpful soul that had drawn our attention to the SSH documentation and to that environment variable.

You're welcome 😉 (even if it has been a pain to find)

@SlugFiller 

>As a test, try running the program directly from inside cmd or bash. It should print whatever you input into it, to the console. It should be as simple as opening cmd, bash, or powershell

I had some pain trying to debug that (because I was suspecting that it was not the wrapper but maybe due to console/shell encoding issue) and found out that **only** bash is outputing something.

So the easiest is maybe redirecting to a file with  `.\GitExtSshAskPass.exe > output.txt` and check file content (if you don't want to use git bash indeed)

**Comment by Binarus** [comment]
At first, thank you very much for the reply.

> Did you test in older versions of Git Extensions, whether the "mini console" even works for your password? Because it's not 100% compatible with `cmd` or `bash`, and tends to mess up input if you need to press any special keys (e.g. ctrl, alt, backspace, etc).

Yes, I was constantly working performing certain tasks in Git bash and other tasks with Git Extensions. From that, I can tell for sure that the "mini console" worked reliably for all my passphrases. The same is true for Git bash (although that was not your question :-)).
 
> Also, looking at it more closely, the issue could actually be password length, rather than contents. While the wrapper has support for unicode (`TCHAR` is used everywhere), the password is [limited to 100 characters](https://github.com/gitextensions/gitextensions/blob/44be238d28e5c68ee11bca3f67351d7383777e65/GitExtSshAskPass/SshAskPass.cpp#L43). Could this be part of the issue?

No, because my passphrases are shorter than 100 characters. They are usually around 50 characters long. Those 50 characters may take more than 100 bytes in memory, depending on the encoding, but that shouldn't be a problem if the limitation is 100 *characters*, not 100 *bytes*.

> As a test, try running the program directly from inside `cmd` or `bash`. It should print whatever you input into it, to the console. It should be as simple as opening `cmd`, `bash`, or `powershell` from the Git Extensions toolbar, and then running:
> 
>     * `bash`: `"$SSH_ASKPASS"`
> 
>     * `cmd`: `"%SSH_ASKPASS%" | more`
> 
>     * `powershell`: `& $env:SSH_ASKPASS | more`
> 
> 
> It should print the result to the console. If the result comes out garbled or truncated, it indicates there's a problem in the wrapper.

Thank you very much for this hint! This clearly shows that there is a problem in the wrapper, as already suspected. I have tried with one of my passphrases and have seen that the following characters get garbled:

`ü ° ä ´` (the last one is a forward tick)

Of course, this list only contains the problematic characters that are in the one passphrase I have tested so far. It seems that the wrapper only processes ASCII characters correctly, or that the conversion from or to the external encoding fails.

Best regards,

Binarus

**Comment by Binarus** [comment]
@pmiossec Thank you very much! This is just to let you know that I'll test your suggestions in a few hours. I'll report my findings today evening.

**Comment by Binarus** [comment]
@pmiossec 

> Can you try this attempt of a fix #11381 ? Get the new `GitExtSshAskPass.exe` from the msi file (the portable version doesn't have it inside 😢 ) in artifacts: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/48585850/artifacts
> 
> Then uncompress the msi and get `GitExtSshAskPass.exe` from there (7-zip is doing it well).
> 
> Before I had '?' or some strange chars in the output and with this fix, I have the values entered in the wrapper. So I hope it will fix the issue.

Thank you very much for that fix! I confirm that it works for me in the following three situations:

- The test described by @SlugFiller (the variant in Git bash): When executing your new GitExtSshAskPass.exe and entering the problematic passphrase as in the first test, the terminal console now echoes the passphrase correctly.
- In Git Extensions, when I choose "Commands -> Pull/Fetch...", adjust the options in the dialog box that appears, hit the "Fetch" button, the command line stdin wrappers appears (same as before as expected), but now, when I type my passphrase, it processes it correctly, Git Extensions unlocks the SSH key and accesses the remote repository as desired.
- The same is true when I work in Git bash (that is, if I reset the environment so that the stdin wrapper appears again when it needs the SSH key passphrase, the wrapper now processes the passphrase correctly).

Of course, for the latter two tests, I have copied `GitExtSsshAskPass.exe` from your new MSI installer over the version in the Git Extensions installation directory.

I am quite sure that you have fixed the bug, although I would need more time to test it with all my passphrases or eventually to test a few new passphrases that contain Kyrillic, Greek, Chinese (and so on) characters.

You really have treated this issue in an exemplary way. Thank you very much for that!

However, may I suggest that you include tests with a few passphrases with exotic Unicode characters in your test suites to avoid regressions in the future (hint: include also characters in those passphrases that are encoded in 4 bytes in UTF-8 (surrogate pairs))?

Thanks again, and best regards,

Binarus
